processing/src/main/java/org/apache/druid/query/operator/GlueingPartitioningOperator.java

```diff
@@ -0,0 +1,279 @@
+      }
+      this.boundaries = groupPartitioner.computeBoundaries(partitionColumns);
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+      return currentIndex < boundaries.length - 1;
+    }
+
+    /**
+     * Retrieves the next partition in the RowsAndColumns. If the first partition has not been handled yet,
+     * it may be glued with the previous RowsAndColumns if the partition columns match.
+     *
+     * @return The next RowsAndColumns partition, potentially glued with the previous one.
+     * @throws NoSuchElementException if there are no more partitions.
+     */
+    @Override
+    public RowsAndColumns next()
+    {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+
+      if (!firstPartitionHandled) {
+        firstPartitionHandled = true;
+        int start = boundaries[currentIndex];
+        int end = boundaries[currentIndex + 1];
+        LimitedRowsAndColumns limitedRAC = new LimitedRowsAndColumns(rac, start, end);
+
+        if (isGlueingNeeded(previousRac, limitedRAC)) {
+          RowsAndColumns gluedRAC = getConcatRacForFirstPartition(previousRac, limitedRAC);
+          ensureMaxRowsMaterializedConstraint(gluedRAC.numRows());
+          previousRac = null;
+          currentIndex++;
+          return gluedRAC;
+        } else {
+          if (previousRac != null) {
+            RowsAndColumns temp = previousRac;
+            previousRac = null;
+            return temp;
+          }
+        }
+      }
+
+      int start = boundaries[currentIndex];
+      int end = boundaries[currentIndex + 1];
+      currentIndex++;
+      return new LimitedRowsAndColumns(rac, start, end);
+    }
+
+    /**
+     * Determines whether gluing is needed between the previous RowsAndColumns and the first partition
+     * of the current RowsAndColumns based on the partition columns. If the columns match, the two RACs
+     * will be glued together.
+     *
+     * @param previousRac The previous RowsAndColumns.
+     * @param firstPartitionOfCurrentRac The first partition of the current RowsAndColumns.
+     * @return true if gluing is needed, false otherwise.
+     */
+    private boolean isGlueingNeeded(RowsAndColumns previousRac, RowsAndColumns firstPartitionOfCurrentRac)
+    {
+      if (previousRac == null) {
+        return false;
+      }
+
+      final ConcatRowsAndColumns concatRac = getConcatRacForFirstPartition(previousRac, firstPartitionOfCurrentRac);
+      for (String column : partitionColumns) {
+        final Column theCol = concatRac.findColumn(column);
+        if (theCol == null) {
+          continue;
+        }
+        final ColumnAccessor accessor = theCol.toAccessor();
+        // Compare 1st row of previousRac and firstPartitionOfCurrentRac in [previousRac, firstPartitionOfCurrentRac] form.
+        int comparison = accessor.compareRows(0, previousRac.numRows());
+        if (comparison != 0) {
+          return false;
+        }
+      }
+      return true;
```

Comment:
`return true` as the default?

Reply:
Yeah, if we didn't encounter any differences in row values for `partition by` columns, then we can glue them together, hence return true.